core/blockchain.go

```diff
@@ -1255,6 +1257,15 @@ func (bc *BlockChain) WriteBlockWithState(
 		return NonStatTy, err
 	}
 
+	if bc.chainConfig.IsEnablePruneBeaconChainFeature() {
+		if block.Number().Cmp(big.NewInt(pruneBeaconChainBlockBefore)) > 0 && block.Epoch().Cmp(big.NewInt(pruneBeaconChainBeforeEpoch)) > 0 {
+			// probability of 1 in 100 to prune
+			maxBlockNum := big.NewInt(0).Sub(block.Number(), big.NewInt(pruneBeaconChainBlockBefore)).Uint64()
+			maxEpoch := big.NewInt(0).Sub(block.Epoch(), big.NewInt(pruneBeaconChainBeforeEpoch))
+			go bc.blockchainPruner.StartPruneBeaconChain(bc.db, maxBlockNum, maxEpoch)
```

Comment:
Looks messy. So there will be multiple goroutines doing the prune at the same time, and the same block looks like will be pruned multiple times... This is not optimal. You should write some scheduler logic for scheduling the pruning tasks.

Reply:
As above, the same block will not be iterated or deleted multiple times, and there is an atomic lock inside, so only one is executing at the same time